// @hono/vite-build/dist/adapter/cloudflare-pages/index.mjs
import buildPlugin, { defaultOptions } from "../../base.mjs";
import { resolve } from "node:path";
import { readdir, writeFile } from "node:fs/promises";
//#region src/adapter/cloudflare-pages/index.ts
const WORKER_JS_NAME = "_worker.js";
const ROUTES_JSON_NAME = "_routes.json";
const cloudflarePagesBuildPlugin = (pluginOptions) => {
	let config;
	const staticPaths = [];
	return {
		...buildPlugin({
			...pluginOptions,
			output: WORKER_JS_NAME
		}),
		configResolved: async (resolvedConfig) => {
			config = resolvedConfig;
		},
		writeBundle: async () => {
			const paths = await readdir(resolve(config.root, config.build.outDir), { withFileTypes: true });
			if (paths.some((p) => p.name === ROUTES_JSON_NAME)) return;
			else {
				paths.forEach((p) => {
					if (p.isDirectory()) staticPaths.push(`/${p.name}/*`);
					else {
						if (p.name === WORKER_JS_NAME) return;
						staticPaths.push(`/${p.name}`);
					}
				});
				const staticRoutes = {
					version: 1,
					include: ["/*"],
					exclude: staticPaths
				};
				await writeFile(resolve(config.root, pluginOptions?.outputDir ?? defaultOptions.outputDir, "_routes.json"), JSON.stringify(staticRoutes));
			}
		},
		name: "@hono/vite-build/cloudflare-pages"
	};
};
//#endregion
export { cloudflarePagesBuildPlugin as default };
