// @hono/vite-build/dist/adapter/bun/index.mjs
import buildPlugin, { defaultOptions as defaultOptions$1 } from "../../base.mjs";
import { serveStaticHook } from "../../entry/serve-static.mjs";
//#region src/adapter/bun/index.ts
const defaultOptions = {
	...defaultOptions$1,
	entryContentAfterHooks: [() => `
      let websocket
      for (const [, app] of Object.entries(modules)) {
        if (
          app &&
          typeof app === 'object' &&
          'websocket' in app &&
          app.websocket !== undefined
        ) {
          if (websocket !== undefined) {
            throw new Error(
              \`Handler "websocket" is defined in multiple entry files. Please ensure each handler is defined only once.\`
            )
          }
          websocket = app.websocket
        }
      }
    `],
	entryContentDefaultExportHook: (appName) => `export default websocket !== undefined ? { fetch: ${appName}.fetch.bind(${appName}), websocket } : ${appName}`
};
const bunBuildPlugin = (pluginOptions) => {
	return {
		...buildPlugin({
			ssrTarget: "node",
			entryContentBeforeHooks: [async (appName, options) => {
				let code = "import { serveStatic } from 'hono/bun'\n";
				code += serveStaticHook(appName, {
					filePaths: options?.staticPaths,
					root: pluginOptions?.staticRoot
				});
				return code;
			}],
			...pluginOptions,
			external: ["bun", ...pluginOptions?.external ?? []],
			entryContentAfterHooks: pluginOptions?.entryContentAfterHooks ?? defaultOptions.entryContentAfterHooks,
			entryContentDefaultExportHook: pluginOptions?.entryContentDefaultExportHook ?? defaultOptions.entryContentDefaultExportHook
		}),
		name: "@hono/vite-build/bun"
	};
};
//#endregion
export { bunBuildPlugin as default, defaultOptions };
