// @hono/vite-build/dist/adapter/node/index.mjs
import buildPlugin from "../../base.mjs";
import { serveStaticHook } from "../../entry/serve-static.mjs";
//#region src/adapter/node/index.ts
const nodeBuildPlugin = (pluginOptions) => {
	const port = pluginOptions?.port ?? 3e3;
	const shutdownTimeoutMs = pluginOptions?.shutdownTimeoutMs;
	return {
		...buildPlugin({
			ssrTarget: "node",
			entryContentBeforeHooks: [async (appName, options) => {
				let code = "import { serveStatic } from '@hono/node-server/serve-static'\n";
				code += serveStaticHook(appName, {
					filePaths: options?.staticPaths,
					root: pluginOptions?.staticRoot
				});
				return code;
			}],
			entryContentAfterHooks: [async (appName) => {
				let code = "import { serve } from '@hono/node-server'\n";
				if (shutdownTimeoutMs !== void 0) {
					code += `const server = serve({ fetch: ${appName}.fetch, port: ${port.toString()} })\n`;
					code += "const gracefulShutdown = () => {\n";
					code += "  server.close(() => process.exit(0))\n";
					if (shutdownTimeoutMs > 0) code += `  setTimeout(() => process.exit(1), ${shutdownTimeoutMs}).unref()\n`;
					code += "}\n";
					code += "process.on('SIGINT', gracefulShutdown)\n";
					code += "process.on('SIGTERM', gracefulShutdown)";
				} else code += `serve({ fetch: ${appName}.fetch, port: ${port.toString()} })`;
				return code;
			}],
			...pluginOptions
		}),
		name: "@hono/vite-build/node"
	};
};
//#endregion
export { nodeBuildPlugin as default };
