// @hono/vite-build/dist/adapter/vercel/index.mjs
import { getEntryContent } from "../../entry/index.mjs";
import { defaultOptions } from "../../base.mjs";
import { builtinModules } from "module";
import { existsSync, mkdirSync, readFileSync } from "node:fs";
import { resolve } from "node:path";
import { cp, writeFile } from "node:fs/promises";
//#region src/adapter/vercel/index.ts
const BUNDLE_NAME = "index.js";
const DEFAULT_FUNCTION_NAME = "__hono";
const VIRTUAL_ENTRY_PREFIX = "virtual:build-entry-module-vercel-";
const functionEntryHooks = {
	entryContentAfterHooks: [() => "import { handle } from '@hono/node-server/vercel'"],
	entryContentDefaultExportHook: (appName) => `export default handle(${appName})`
};
const configWriteQueues = /* @__PURE__ */ new Map();
const writeJSON = (path, data) => {
	const dir = resolve(path, "..");
	if (!existsSync(dir)) mkdirSync(dir, { recursive: true });
	return writeFile(path, JSON.stringify(data));
};
const readJSON = (path) => {
	if (!existsSync(path)) return;
	return JSON.parse(readFileSync(path, "utf-8"));
};
const enqueueConfigWrite = async (key, writeTask) => {
	const previousTask = configWriteQueues.get(key);
	if (previousTask) await previousTask;
	const nextTask = writeTask();
	configWriteQueues.set(key, nextTask);
	try {
		await nextTask;
	} finally {
		if (configWriteQueues.get(key) === nextTask) configWriteQueues.delete(key);
	}
};
const getRuntimeVersion = () => {
	try {
		const systemNodeVersion = process.versions.node.split(".")[0];
		return `nodejs${Number(systemNodeVersion)}.x`;
	} catch {
		return "nodejs22.x";
	}
};
const escapeRouteSegment = (value) => {
	return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
};
const getDefaultRoutePattern = (functionName) => {
	return `^/${escapeRouteSegment(functionName)}(?:/.*)?$`;
};
const appendRouteIfMissing = (target, seen, route) => {
	const key = JSON.stringify(route);
	if (seen.has(key)) return;
	seen.add(key);
	target.push(route);
};
const getFunctionConfig = (config, functionConfig) => {
	return {
		...functionConfig,
		runtime: getRuntimeVersion(),
		launcherType: "Nodejs",
		handler: BUNDLE_NAME,
		shouldAddHelpers: Boolean(functionConfig?.shouldAddHelpers),
		shouldAddSourcemapSupport: Boolean(config.build.sourcemap),
		supportsResponseStreaming: true
	};
};
const getRoutesForFunction = (functionName, configuredRoutes) => {
	if (configuredRoutes && configuredRoutes.length > 0) return configuredRoutes;
	if (functionName === DEFAULT_FUNCTION_NAME) return [{
		src: "/(.*)",
		dest: `/${DEFAULT_FUNCTION_NAME}`
	}];
	return [{
		src: getDefaultRoutePattern(functionName),
		dest: `/${functionName}`
	}];
};
const mergeVercelConfig = async (configPath, routesToAdd, configOverride, functionName) => {
	const existingConfig = readJSON(configPath);
	const mergedRoutes = [];
	const seenRoutes = /* @__PURE__ */ new Set();
	for (const route of existingConfig?.routes ?? []) appendRouteIfMissing(mergedRoutes, seenRoutes, route);
	for (const route of configOverride?.routes ?? []) appendRouteIfMissing(mergedRoutes, seenRoutes, route);
	appendRouteIfMissing(mergedRoutes, seenRoutes, { handle: "filesystem" });
	for (const route of routesToAdd) appendRouteIfMissing(mergedRoutes, seenRoutes, {
		...route,
		dest: route.dest ?? `/${functionName}`
	});
	await writeJSON(configPath, {
		...existingConfig,
		...configOverride,
		version: 3,
		routes: mergedRoutes
	});
};
const copyStaticFiles = async (publicDirPath, outputDir) => {
	if (!existsSync(publicDirPath)) return;
	try {
		await cp(publicDirPath, resolve(outputDir, "static"), {
			recursive: true,
			force: true
		});
	} catch (error) {
		if (error.code !== "EEXIST") throw error;
	}
};
const vercelBuildPlugin = (pluginOptions) => {
	let config;
	const functionName = pluginOptions?.vercel?.name ?? DEFAULT_FUNCTION_NAME;
	if (!functionName) throw new Error("`vercel.name` is required and cannot be empty.");
	const virtualEntryId = `${VIRTUAL_ENTRY_PREFIX}${functionName}`;
	const resolvedVirtualEntryId = `\0${virtualEntryId}`;
	return {
		name: "@hono/vite-build/vercel",
		apply: pluginOptions?.apply ?? defaultOptions.apply,
		resolveId(id) {
			if (id === virtualEntryId) return resolvedVirtualEntryId;
		},
		async load(id) {
			if (id !== resolvedVirtualEntryId) return;
			const entry = pluginOptions?.entry ?? defaultOptions.entry;
			return await getEntryContent({
				entry: Array.isArray(entry) ? entry : [entry],
				entryContentBeforeHooks: pluginOptions?.entryContentBeforeHooks,
				entryContentAfterHooks: pluginOptions?.entryContentAfterHooks ?? functionEntryHooks.entryContentAfterHooks,
				entryContentDefaultExportHook: pluginOptions?.entryContentDefaultExportHook ?? functionEntryHooks.entryContentDefaultExportHook,
				staticPaths: pluginOptions?.staticPaths,
				preset: pluginOptions?.preset
			});
		},
		configResolved: (resolvedConfig) => {
			config = resolvedConfig;
		},
		config: async () => {
			return {
				ssr: {
					external: pluginOptions?.external ?? defaultOptions.external,
					noExternal: true,
					target: "node"
				},
				build: {
					outDir: ".vercel/output",
					emptyOutDir: pluginOptions?.emptyOutDir ?? defaultOptions.emptyOutDir,
					minify: pluginOptions?.minify ?? defaultOptions.minify,
					ssr: true,
					rollupOptions: {
						external: [...builtinModules, /^node:/],
						input: { [functionName]: virtualEntryId },
						output: { entryFileNames: `functions/[name].func/${BUNDLE_NAME}` }
					}
				}
			};
		},
		writeBundle: async () => {
			const outputDir = resolve(config.root, config.build.outDir);
			const functionDir = resolve(outputDir, "functions", `${functionName}.func`);
			const configPath = resolve(outputDir, "config.json");
			const publicDirPath = resolve(config.root, config.publicDir);
			const routesToAdd = getRoutesForFunction(functionName, pluginOptions?.vercel?.routes);
			const functionConfig = getFunctionConfig(config, pluginOptions?.vercel?.function);
			await copyStaticFiles(publicDirPath, outputDir);
			await Promise.all([writeJSON(resolve(functionDir, ".vc-config.json"), functionConfig), writeJSON(resolve(functionDir, "package.json"), { type: "module" })]);
			await enqueueConfigWrite(configPath, async () => {
				await mergeVercelConfig(configPath, routesToAdd, pluginOptions?.vercel?.config, functionName);
			});
		}
	};
};
//#endregion
export { vercelBuildPlugin as default };
